client/snapshot.go

```diff
@@ -0,0 +1,151 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// errors on reading:
+var (
+	ErrSnapshotUnsupported  = errors.New("unsupported snapshot format")
+	ErrSnapshotSizeMismatch = errors.New("reported size and read size differ")
+	ErrSnapshotHashMismatch = errors.New("hash mismatch")
+)
+
+// errors about not finding stuff (should bubble out as something more user-friendly, probably)
+var (
+	ErrSnapshotNotFound      = errors.New("no snapshot by that ID")
+	ErrSnapshotSnapsNotFound = errors.New("the snapshot with the given ID does not contain the requested snaps")
+)
+
+// A SnapshotOp is used to request an operation on a snapshot
+type SnapshotOp struct {
+	ID     uint64   `json:"ID"`
+	Action string   `json:"action"`
+	Snaps  []string `json:"snaps,omitempty"`
+	Homes  []string `json:"snaps,omitempty"`
+}
+
+func (op *SnapshotOp) String() string {
+	// verb of snapshot #N [for snaps %q] [for homes %q]
+	var snaps string
+	var homes string
+	if len(op.Snaps) > 0 {
+		snaps = " for snaps " + strutil.Quoted(op.Snaps)
+	}
+	if len(op.Homes) > 0 {
+		homes = " for homes " + strutil.Quoted(op.Homes)
+	}
+	return fmt.Sprintf("%s of snapshot #%d%s%s", strings.Title(op.Action), op.ID, snaps, homes)
+}
+
+// A Snapshot is a collection of archives with a simple metadata json file
+// (and hashsums of everything)
+//
+// XXX can snapd sign these?
+type Snapshot struct {
+	// ID of the snapshot group (a snapshot group is a "snap save" invocation)
+	ID uint64 `json:"snapshot"`
```

Comment:
If I understand what it implies, the name feels misleading, as we'd expect uniqueness. I'd expect an ID to not repeat. "Group" or "Sequence" might be better names.

Reply:
yeah, should be `Group` or `GroupID`.